src/middleware/aws/instrumentation/stream.rs

```diff
@@ -0,0 +1,221 @@
+///
+/// All instrumented streams automatically include the `aws.pagination_stream = true`
+/// attribute to help identify streaming operations in traces.
+///
+/// # Examples
+///
+/// ```rust
+/// use aws_sdk_dynamodb::{Client as DynamoClient, types::AttributeValue};
+/// use futures_util::TryStreamExt;
+/// use telemetry_rust::{
+///     KeyValue,
+///     middleware::aws::{AwsStreamInstrument, DynamodbSpanBuilder},
+///     semconv,
+/// };
+///
+/// async fn query_table() -> Result<usize, Box<dyn std::error::Error>> {
+///     let config = aws_config::load_from_env().await;
+///     let dynamo_client = DynamoClient::new(&config);
+///     let items =
+///         dynamo_client
+///             .query()
+///             .table_name("table_name")
+///             .index_name("my_index")
+///             .key_condition_expression("PK = :pk")
+///             .expression_attribute_values(":pk", AttributeValue::S("Test".to_string()))
+///             .into_paginator()
+///             .items()
+///             .send()
+///             .instrument(DynamodbSpanBuilder::query("table_name").attribute(
+///                 KeyValue::new(semconv::AWS_DYNAMODB_INDEX_NAME, "my_index"),
+///             ))
+///             .try_collect::<Vec<_>>()
+///             .await?;
+///     println!("DynamoDB items: {items:#?}");
+///     Ok(items.len())
+/// }
+/// ```
+pub trait AwsStreamInstrument<T, E, S>
+where
+    E: RequestId + Error,
+    S: Stream<Item = Result<T, E>>,
+{
+    /// Instruments the stream with OpenTelemetry tracing.
+    ///
+    /// This method wraps the stream in an [`InstrumentedStream`] that will:
+    /// - Start a span when the stream begins polling
+    /// - End the span with success when the stream completes normally
+    /// - End the span with error information if the stream encounters an error
+    ///
+    /// # Arguments
+    ///
+    /// * `span` - The span builder or span configuration to use for instrumentation
+    ///
+    /// # Returns
+    ///
+    /// An [`InstrumentedStream`] that wraps the original stream with telemetry capabilities.
+    fn instrument<'a>(
+        self,
+        span: impl Into<AwsSpanBuilder<'a>>,
+    ) -> InstrumentedStream<'a, S>;
+}
+
+impl<T, E, S> AwsStreamInstrument<T, E, S> for S
+where
+    E: RequestId + Error,
+    S: Stream<Item = Result<T, E>>,
+{
+    fn instrument<'a>(
+        self,
+        span: impl Into<AwsSpanBuilder<'a>>,
+    ) -> InstrumentedStream<'a, S> {
+        InstrumentedStream {
+            inner: self,
+            state: Cell::new(StreamState::new(span)),
+        }
+    }
+}
+
+impl<T, E> AwsStreamInstrument<T, E, PaginationStreamImplStream<Result<T, E>>>
+    for PaginationStream<Result<T, E>>
+where
+    E: RequestId + Error,
+{
+    fn instrument<'a>(
+        self,
+        span: impl Into<AwsSpanBuilder<'a>>,
+    ) -> InstrumentedStream<'a, PaginationStreamImplStream<Result<T, E>>> {
+        self.into_stream_03x().instrument(span)
```

Comment:
AWS `PaginationStream` is a custom stream that doesn't implement `Stream` trait out of the box.

I guess I could've created a "similar" stream with exactly the same methods, but it seems like an over-complication.

`into_stream_03x` from `aws_smithy_types_convert` converts custom `PaginationStream` into a normal `Stream`, so I just made an instrumentation for that.

I ended up with very generic `InstrumentedStream` which works for any `TryStream` that return an AWS error and this shorthand for `PaginationStream`.